### PR TITLE
L2 Regularization for all layers

### PR DESCRIPTION
**Pull request by dust0x**
It is better practice to apply weight decay methods in each layer. This commit proposes to change the current implementation of L2 Regularization from just the output layer to the hidden (convolution) layers and also omits regularization of the bias at the output as an unnecessary overhead.

Seeking review and testing.

### CODE REVIEW

File: text_cnn.py [hunk 1]
```diff
@@ -68,11 +65,12 @@ def __init__(
                 shape=[num_filters_total, num_classes],
                 initializer=tf.contrib.layers.xavier_initializer())
             b = tf.Variable(tf.constant(0.1, shape=[num_classes]), name="b")
-            l2_loss += tf.nn.l2_loss(W)
-            l2_loss += tf.nn.l2_loss(b)
             self.scores = tf.nn.xw_plus_b(self.h_drop, W, b, name="scores")
             self.predictions = tf.argmax(self.scores, 1, name="predictions")
 
+        # Calculate L2 Regularization
+        l2_loss = tf.add_n([tf.nn.l2_loss(v) for v in tf.trainable_variables() if "b" not in v.name])
```
Comment (passerbythesun) [inline text_cnn.py]: This will also add embedding weights into l2_loss, is it intended?

Comment (dust0x) [reply text_cnn.py]: It isn't! Thanks for pointing this out.

Following change should fix it I think:

`l2_loss = tf.add_n([tf.nn.l2_loss(v) for v in tf.trainable_variables()[1:] if "W" in v.name])`

**Comment by dennybritz** [comment]
Thanks. This looks good in principle, have you tested it?

**Comment by dust0x** [comment]
@dennybritz: I have only tested it on a small subset of the data. The program itself runs smoothly. However, I have not been able to do any meaningful performance/accuracy evaluations against your implementation because of a resource constraint. I _can_ make a few citations to justify that this method works better in theory.

[This](https://stackoverflow.com/a/38466108) StackOverflow answer verifies that the implementation is indeed correct and something of a standard, as similar usage can be found elsewhere.

Discussion on (not) regularizing bias can be found [here](http://deeplearning.stanford.edu/wiki/index.php/Backpropagation_Algorithm) and in [this](http://www.deeplearningbook.org/contents/regularization.html) chapter of the Deep Learning book (p. 226).

**Comment by chiragnagpal** [comment]
Weight Decay with Adam is not a good idea. You'll end up adding another hyperparameter with little performance gain.

**Comment by dust0x** [comment]
@chiragnagpal yes, weight decay is largely ineffective when used with Adam. I'm only suggesting a modification on what is currently implemented. 

We could possibly rethink the entire implementation as proposed in [Fixing Weight Decay Regularization in Adam](https://arxiv.org/pdf/1711.05101.pdf) (Loshchilov, Hutter; 2018).

**Comment by chiragnagpal** [comment]
That paper is too new with hardly any citations. 

1) It would be unwise to change from a standard implementation to a more esoteric one, especially given that we do not have any numbers on the performance improvement. 

2) Tensorflow itself may add weight decay to the optimizers, like pytorch and dynet do. Why bother with doing it explicitly just now, given that it might break again in the future.

**Comment by dust0x** [comment]
Sure. It's only something to think about.
